Add kmigrator DB level constraints (#524)
open-condo-software/condo: docs/kmigrator.md

# kmigrator

> It's CLI migration tool. It helps you to create a `migrations` folder.

Author: ?

Example:

```
const User = new GQLListSchema('User', {
    schemaDoc: 'Individual / person / service account / impersonal company account',
    fields: {
        ...
        email: {
            schemaDoc: 'Email. Transformed to lower case',
            type: Text,
            kmigratorOptions: { null: false, unique: true },
        },
        birth: {
            type: DateTimeUtc,
            kmigratorOptions: { db_index: true },
        },
        ...
    },
})
```

#### on_delete

It set relations on delete behavior.

For example, if you have foregin key from Ticket.assignee to User.id you need to set 
on delete User behaviour for Ticket.assignee column. 
 
You can use options:

 - `models.PROTECT` -- forbid the deletion of the referenced object. To delete it you will have to delete all objects that reference it manually. SQL equivalent: `RESTRICT`.
 - `models.SET_NULL` -- set the reference to NULL (requires the field to be nullable). For instance, when you delete a User, you might want to keep the comments he posted on blog posts, but say it was posted by an anonymous (or deleted) user. SQL equivalent: SET NULL.

>  - `models.CASCADE` -- when the referenced object is deleted, also delete the objects that have references to it (when you remove a blog post for instance, you might want to delete comments as well). SQL equivalent: CASCADE.

Reviewer: ```suggestion
 - `models.CASCADE` deletes the object on delete of referenced object. (when you remove a blog post for instance, you might want to delete comments as well). SQL equivalent: `CASCADE`.
```

## kmigrator table options

you can set options: 
 - `constraints` -- to create database level constrains
 - `indexes` -- to create database level indexes

#### constraints

Example:

```
const User = new GQLListSchema('User', {
    schemaDoc: 'Individual / person who have some access to service',
    fields: {
        type: {
            schemaDoc: 'We have b2b and b2c user segments',
            type: Select,
            dataType: 'enum',
            options: ['b2b', 'b2c', 'sys'],
            defaultValue: 'b2b',
            isRequired: true,
        },
        phone: {
            schemaDoc: 'Phone. In international E.164 format without spaces',
            type: Text,
            isRequired: true,
        },
        ...
    },
    kmigratorOptions: {
        constraints: [
            {

>                 type: 'models.UniqueConstraint',

Reviewer: Can we simplify interface for developer and not expose internal kmigrator conventions?
```suggestion
                type: 'unique',
```

Author: I don't think that it's a good idea. Because it uses `models.UniqueConstraint` under the hood. 

It's better to do it by another task. We also need to move from the `django` `Q` object syntax too in this task.

                fields: ['type', 'phone'],
                name: 'unique_type_and_phone',
            },
        ],
    },
})
```

> This example generate SQL like so: `ALTER TABLE "User" ADD CONSTRAINT "unique_type_and_phone" UNIQUE ("type", "phone");`

Reviewer: ```suggestion
This example generate following SQL: `ALTER TABLE "User" ADD CONSTRAINT "unique_type_and_phone" UNIQUE ("type", "phone");`
```

Another example:
```
const User = new GQLListSchema('User', {
    schemaDoc: 'Individual / person who have some access to service',
    fields: {
        phone: {
            schemaDoc: 'Phone. In international E.164 format without spaces',
            type: Text,
            isRequired: false,
        },
        email: {
            schemaDoc: 'Email',
            type: Text,
            isRequired: false,
        },
        ...
    },
    kmigratorOptions: {
        constraints: [
            {
                type: 'models.CheckConstraint',
                check: 'Q(phone__isnull=False) | Q(email__isnull=False)',
                name: 'has_phone_or_email',
            },
        ],
    },
})
```

> This example generate SQL like so: `ALTER TABLE "User" ADD CONSTRAINT "has_phone_or_email" CHECK (("phone" IS NOT NULL OR "email" IS NOT NULL));`

Reviewer: ```suggestion
This example generates following SQL: `ALTER TABLE "User" ADD CONSTRAINT "has_phone_or_email" CHECK (("phone" IS NOT NULL OR "email" IS NOT NULL));`
```

You can also define a `condition` for `models.UniqueConstraint`:

```
{
    type: 'models.UniqueConstraint',
    fields: ['type', 'phone'],
    condition: 'Q(deletedAt__isnull=True)',
    name: 'unique_type_and_phone',
},
```

SQL result: `CREATE UNIQUE INDEX "unique_type_and_phone" ON "User" ("type", "phone") WHERE "deletedAt" IS NULL;`

Another example:
```
{

>     type: 'models.CheckConstraint',

Reviewer: Can we simplify interface for developer and not expose internal kmigrator conventions?

```suggestion
    type: 'check',
```
